components/fxa-client/src/lib.rs

```diff
@@ -203,13 +233,47 @@ impl FirefoxAccount {
             None => Err(ErrorKind::NoRefreshToken.into()),
         }
     }
+
+    /// Disconnect from the account and optionaly destroy our device record. This will
+    /// leave the account object in a state where it can eventually reconnect to the same user.
+    /// This is a "best effort" infallible method: e.g. if the network is unreachable,
+    /// the device could still be in the FxA devices manager.
+    ///
+    /// **💾 This method alters the persisted account state.**
+    pub fn disconnect(&mut self) {
+        // Save the refresh token before resetting the state.
+        let refresh_token = match self.state.refresh_token {
+            Some(ref t) => Some(t.token.clone()),
+            None => None,
+        };
+        self.start_over();
+        if let Some(refresh_token) = refresh_token {
+            // Delete the current device (which deletes the refresh token), or
+            // the refresh token directly if we don't have a device.
+            let destroy_result = match self.get_current_device() {
+                // If we get an error trying to fetch our device record we'll at least
+                // still try to delete the refresh token itself.
+                Ok(Some(device)) => {
+                    self.client
+                        .destroy_device(&self.state.config, &refresh_token, &device.id)
+                }
+                _ => self
+                    .client
+                    .destroy_oauth_token(&self.state.config, &refresh_token),
```

Comment:
Ugh, I just realized that this won't work. It will attempt to destroy the `refresh_token` as though it were an `access_token`, which will fail. We'll need to add a separate `destroy_refresh_token` method that calls the same endpoint on the oauth-server, but which specifies `refresh_token` in the request body rather than `token`. Bleh :-(

Reply:
oh oh :/
https://github.com/mozilla/application-services/blob/e43222df4972a9aad535f2cc5bc67e4f8e19f035/components/fxa-client/src/oauth.rs#L231-L240

Reply:
Yikes. Oh well, at least we found it! :-)